lib/uri/common.rb

```diff
@@ -62,10 +65,26 @@ def make_components_hash(klass, array_hash)
 
   include REGEXP
 
-  @@schemes = {}
+  SCHEMES = {}.freeze
+  private_constant :SCHEMES
+
   # Returns a Hash of the defined schemes.
   def self.scheme_list
-    @@schemes
+    SCHEMES
+  end
+
+  # Registers a new scheme when adding custom URIs.
+  # Example:
+  #   module URI
+  #     class RSYNC < Generic
+  #       DEFAULT_PORT = 873
+  #     end
+  #     register_scheme('RSYNC', RSYNC)
+  #   end
+  def self.register_scheme(name, mod)
+    updated = SCHEMES.merge(name.to_s => mod).freeze
+    remove_const(:SCHEMES)
+    const_set(:SCHEMES, updated)
```

Comment:
This seems quite hacky and is extremely inefficient as well: for every call to `register_scheme` you will typically invalidate all constant lookup caches under `URI`, and potentially cause deoptimizations for JITed code.
Also the fact it's exposed as a public API seems a huge footgun given this drawback.

There must be a better way to handle this.
If there isn't, then something should be improved in Ractor.

For instance maybe having `SCHEMES` initially mutable and at the end of `lib/uri.rb` make it shareable?
I guess that doesn't work for "Adding custom URIs" though.

I think using an instance variable on `URI` would work.

Reply:
Also it's obviously not thread-safe and could lead to errors like `no constant URI::SCHEMES`.

Reply:
These are great points. I agree that we might need to come up with something better.

To me this problem sounds generic to all Ruby apps: there's a boot phase when some things might get mutated, and there's a runtime phase where multiple threads/ractors will run where we want all global state to be frozen.

It would be ideal if we could somehow make that a pattern, and then we could freeze `SCHEMES` only once the code is fully loaded and everyone is done adding custom URIs.

Reply:
> for every call to `register_scheme`

which is going to be called a half dozen of times when loading the library...

> the fact it's exposed as a public API

why is that an issue though? It looks to me as it is a feature that there's a API for this, although it need to be documented.

If it's not to be public, then it could be made private.

> Also it's obviously not thread-safe

Right. That should be in the docs. I don't see this as an issue, do not change the configuration after your multi-thread/ractor app has started.

> I think using an instance variable on URI would work.

It won't. You can't access instance variables from shareable objects (like Module or Ractor)

```
ruby -e 'String.singleton_class.attr_accessor :foo; String.foo = 42; Ractor.new { p String.foo }.take'
# => IsolationError
```

Reply:
> It would be ideal if we could somehow make that a pattern

I'm planning on having a small `ractor-config` gem to do this but using the same trick of using `const_set` (with the same caveats, i.e. config only at load time coz it's not thread/ractor safe).

For anything that requires being mutable and shared between ractors, there is no other choice than to involve ractor communication. My gem `ractor-server` makes it easy, but it seems way overkill. A huge issue is also that all apps are currently single-ractor and that the first time another Ractor is started it changes things (I think it lowers performance too).

Reply:
I think it's clear to everyone this is pretty ugly hack.
And IMHO it has no place in any stdlib or gem.
Rather, I think it shows a general problem of Ractor programming being too limiting, for no fundamental reason.
And I agree creating an extra Ractor for this is complete overkill (in terms of code, complexity, extra native thread, extra footprint).

@ko1 What do you think? This is a hack, yes?
There must be a better way.

What about allowing to read/write instance variables of modules, as long as the value is shareable?
Then it would be similar to constants, but it would actually make sense to change it multiple times.
And it would be thread-safe (no time where the ivar is not set except initially).

Another idea is we could have a hook just before the first non-main Ractor is started.
Then you could do something like `Ractor.on_first_ractor { Ractor.make_shareable(URI::SCHEMES) }`.
But I'm not sure it would work well, e.g., if some code creates a Ractor while code is still loading (and that code might be autoloaded so the order is not predictable, and even if loaded explicitly, gems are often loaded "in the middle of loading the app" which could still be problematic).

Reply:
> which is going to be called a half dozen of times when loading the library...

Yes, maybe not a performance issue in this specific case, but the pattern clearly could be.
Do you know any other constant being redefined 6 times?
I think the more important point is redefining constants many times does not make sense conceptually.
It goes against the concept of constants itself.

Reply:
> What about allowing to read/write instance variables of modules, as long as the value is shareable

That would be really nice 👍

Reply:
@marcandre Could file a issue on Redmine to suggest supporting reading/writing `@ivars` of Module? I'll support it.

Reply:
Done! https://bugs.ruby-lang.org/issues/17592
I've asked for read-only, I'll let you argue for read-write if it feels best to you.